portal/models/intervention.py

```diff
@@ -157,6 +157,46 @@ def display_for_user(self, user):
         return DisplayDetails(access=access, intervention=self,
                 user_intervention=ui)
 
+    def quick_access_check(self, user):
+        """Return boolean representing given user's access to intervention
+
+        Somewhat complicated method, depending on intervention configuration.
+        The following ordered steps are used to determine if a user
+        should have access to an intervention.  The first 'true' found
+        is returned (as to make the check as quick as possible).
+
+        1. call each strategy_function in intervention.access_strategies.
+           Note, on rare occasions, a strategy may alter the UserIntervention
+           attributes given the circumstances.
+        2. check for a UserIntervention row defining access for the given
+           user on this intervention.
+        3. check if the intervention has `public_access` set
+
+        @return boolean representing 'access'.
+
+        """
+        # 1. check strategies for access
+        for func in self.fetch_strategies():
+            if func.__name__ == 'update_user_card_html':
+                return True
+            if func(intervention=self, user=user):
+                return True
+
+        # 2. check user_intervention for access
+        ui = UserIntervention.query.filter_by(
+            user_id=user.id, intervention_id=self.id).first()
+        if ui and ui.access == 'granted':
+            return True
+
+        # 3. check intervention scope for access
```

Comment:
@OptimusRhine I *believe* we can run this first for the quick check.  You'd need to review what we do alter in the access strategies - but IIRC that only alters things like the html displayed in the card - doesn't alter access itself?

Reply:
@pbugni I think you're right, it doesn't look like any of the strategies modify `public_access`.

It also looks like we could reverse the order here entirely (i.e. first check for `public_access`, _then_ check the `UserInterventions`, _then_ check the access strategies), unless I'm missing something. That would definitely speed things up...

Reply:
I'm sure I thought there was good reason for the original order, but if you feel you've adequately confirmed none of the current strategies would *need* to run before the other checks, lets change it.

Please double check - I'm sure I only chose this more inefficient order as it was necessary at the time.

Reply:
Well, the order actually still makes since for `display_for_user()`:
* The InterventionStrategy loop needs to happen:
  * ...before the UserIntervention query, because the UserIntervention is returned as part of the `DisplayDetails` object. So if a UI is generated in the IS loop, it needs to happen first.
  * ...no matter what, because of the potential for HTML to be generated, etc
* Given the Strategy loop always needs to happen, the `public_access` check placement doesn't really matter, as long as its there - moving it won't change the runtime of the function.

But none of these are concerns for the quick access check, so I'm going to go ahead and switch that order up, see if it helps things.

Reply:
(switching it up improved things by ~1min)